### PR TITLE
Added various algorithms in python

### PR DESCRIPTION
**Pull request by nchhabra1311**
(Put an `X` inside the `[ ]` to denote check mark `[X]`.)

- [ ] **If creating a new file :**
  - [ ] added links to it in the README files ?
  - [ ] included tests with it ?
  - [ ] added description (overview of algorithm, time and space compleixty, and possible edge case) in docstrings ?

- [ ] **if done some changes :**
  - [ ] wrote short description in the PR explaining what the changes do ?
  - [ ] Fixes #[issue number] if related to any issue

- [ ] **other**

### CODE REVIEW

**Comment by gulshan30310** [comment]
You work is good
If any help required , I can help

File: algorithms/bfs/Bfs.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+from collections import defaultdict
+
+class Graph:
+ 
+    def __init__(self):
+        self.graph = defaultdict(list)
+ 
+    def addEdge(self,u,v):
```
Comment (keon) [inline algorithms/bfs/Bfs.py]: please follow snake_case style

File: algorithms/bfs/Bfs.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        self.graph[u].append(v)
+ 
+    def BFS(self, s):
```
Comment (keon) [inline algorithms/bfs/Bfs.py]: 1. BFS -> bfs
2. we need more descriptive parameter. even better if you could use type hinting.
ex) 
```python
def bfs(self, s: string):
```

File: algorithms/bfs/Bfs.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        visited = [False]*(len(self.graph))
+        queue = []
+        queue.append(s)
+        visited[s] = True
+ 
+        while queue:
+            s = queue.pop(0)
+            print s,
```
Comment (keon) [inline algorithms/bfs/Bfs.py]: we are mainly using python3.
please use `print()`

File: algorithms/bfs/Bfs.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+                    visited[i] = True
+
+    def dfs_visited(self,v,visited):
+        visited[v]= True
+        print v,
+        for i in self.graph[v]:
+            if visited[i] == False:
+                self.dfs_visited(i, visited)
+
+    def DFS(self,v):
```
Comment (keon) [inline algorithms/bfs/Bfs.py]: dfs

File: algorithms/bfs/Bfs.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
+from collections import defaultdict
+
+class Graph:
```
Comment (keon) [inline algorithms/bfs/Bfs.py]: let's move this to graph folder.
I think we can merge this with [this](https://github.com/keon/algorithms/blob/master/algorithms/graph/graph.py#L89)

File: algorithms/bfs/Bfs.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        for i in self.graph[v]:
+            if visited[i] == False:
+                self.dfs_visited(i, visited)
+
+    def DFS(self,v):
+        visited = [False]*(len(self.graph))
+        self.dfs_visited(v,visited)
+        
+g = Graph()
+g.addEdge(0, 3)
```
Comment (keon) [inline algorithms/bfs/Bfs.py]: please move unit test to tests folder

File: algorithms/matrix/matrix chain recurssion.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
+from sys import maxint
```
Comment (keon) [inline algorithms/matrix/matrix chain recurssion.py]: do not use white space in a file name.
`chain_recursion`

File: algorithms/matrix/matrix chain recurssion.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+import time
+def matrix(p,i,j):
+    if i==j:
+        return 0
+    min=maxint
```
Comment (keon) [inline algorithms/matrix/matrix chain recurssion.py]: do not use overlapping variable name with Python native functions. maybe rename to `minimum`?

File: algorithms/matrix/matrix chain recurssion.py [hunk 9]
```diff
@@ -0,0 +1,17 @@
+from sys import maxint
+import time
```
Comment (keon) [inline algorithms/matrix/matrix chain recurssion.py]: 2 new lines after importing statements

File: algorithms/matrix/matrix chain recurssion.py [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+def matrix(p,i,j):
```
Comment (keon) [inline algorithms/matrix/matrix chain recurssion.py]: please use more more descriptive function name.
I don't know what this function is for

File: algorithms/matrix/matrix chain recurssion.py [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+    if i==j:
+        return 0
+    min=maxint
+    for k in range(i,j):
+        count= matrix(p,i,k)+matrix(p,k+1,j)+p[i-1]*p[k]*p[j]
+        if count < min:
+            min=count
+    return min
+
+mat_size=[1,2,3,4,3]
```
Comment (keon) [inline algorithms/matrix/matrix chain recurssion.py]: to unit test

File: algorithms/matrix/matrix chain.py [hunk 12]
```diff
@@ -0,0 +1,23 @@
+from sys import maxint
+import time
```
Comment (keon) [inline algorithms/matrix/matrix chain.py]: i think this is duplicate

File: algorithms/search/SEARCH all.PY [hunk 13]
```diff
@@ -0,0 +1,78 @@
+import time
+global l,l_unsorted
```
Comment (keon) [inline algorithms/search/SEARCH all.PY]: refrain from using global, since we are becoming a library.

File: algorithms/search/SEARCH all.PY [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
... 7957, 7961, 7962, 7966, 7976, 7984, 7986, 7991, 8001, 8020, 8026, 8037, 8092, 8114, 8117, 8122, 8125, 8136, 8157, 8172, 8190, 8213, 8222, 8222, 8223, 8227, 8232, 8241, 8244, 8274, 8287, 8290, 8297, 8300, 8301, 8313, 8314, 8317, 8340, 8340, 8356, 8358, 8359, 8360, 8361, 8371, 8377, 8377, 8384, 8390, 8414, 8423, 8427, 8440, 8447, 8457, 8461, 8485, 8488, 8518, 8519, 8538, 8554, 8566, 8582, 8592, 8626, 8631, 8650, 8664, 8694, 8697, 8712, 8741, 8747, 8754, 8766, 8775, 8790, 8791, 8799, 8831, 8873, 8873, 8885, 8898, 8905, 8908, 8910, 8915, 8919, 8919, 8919, 8932, 8948, 8950, 8960, 8970, 8977, 9003, 9006, 9017, 9019, 9019, 9024, 9042, 9073, 9083, 9084, 9085, 9097, 9101, 9101, 9141, 9160, 9160, 9161, 9166, 9170, 9181, 9187, 9203, 9204, 9212, 9215, 9216, 9219, 9259, 9269, 9274, 9278, 9289, 9301, 9301, 9312, 9317, 9319, 9332, 9362, 9370, 9400, 9406, 9408, 9412, 9415, 9461, 9464, 9478, 9485, 9495, 9498, 9523, 9536, 9536, 9545, 9554, 9580, 9610, 9611, 9616, 9624, 9626, 9630, 9658, 9663, 9708, 9710, 9711, 9721, 9729, 9737, 9752, 9758, 9768, 9773, 9791, 9813, 9818, 9822, 9835, 9836, 9836, 9849, 9855, 9862, 9866, 9873, 9877, 9884, 9890, 9897, 9905, 9909, 9918, 9925, 9936, 9949, 9974, 9975, 9983]
```
Comment (keon) [inline algorithms/search/SEARCH all.PY]: can we simplify this? or have a separate file for this fixture?

File: algorithms/search/SEARCH all.PY [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+        flag=0
+        for i in range(len(l_unsorted)):
+            if l_unsorted[i]==find:
+                print "found at "+str(i+1)
+                flag=1
+                break
+        if flag==0:
+            print "Not Found"
+        print "Time taken is "+str(time.clock()-a)
+    elif choice==2:
```
Comment (keon) [inline algorithms/search/SEARCH all.PY]: I think we can separate each `choice` to a different function.

File: algorithms/strings/kmp.py [hunk 16]
```diff
@@ -0,0 +1,38 @@
+import time
+class KMP:
```
Comment (keon) [inline algorithms/strings/kmp.py]: please add a little more description with big O

File: algorithms/strings/kmp.py [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+        
+        for i in range(len(txt)):
+            while j > 0 and txt[i] != find[j]:
+                j = part[j - 1]
+            if txt[i] == find[j]:
+                j += 1
+            if j == len(find): 
+                ret.append(i - (j - 1))
+                j = 0
+        if ret:
```
Comment (keon) [inline algorithms/strings/kmp.py]: in `search`, let's just return `ret` and let user take care of printing

File: algorithms/strings/kmp.py [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+            print "String found at "
+            for i in ret:
+                print i,
+        else:
+            print "not found"
+    
+obj=KMP()
```
Comment (keon) [inline algorithms/strings/kmp.py]: same. move this unit test to a different testing file.

**Comment by keon** [comment]
Closing as stale — the repo was completely restructured in v1.0.0. Thank you for the contribution\!
